Dockerfile to test building on ubuntu:rolling

LynxAbraxas opened the pull request:
Since #417 is solved with https://github.com/civctp2/civctp2/commit/4c412e6465cb27003d16182794f2c0f817067681 (see https://github.com/civctp2/civctp2/issues/417#issuecomment-2559334624), this PR offers a dockerfile that tests building on ubuntu:rolling, which is at least working for ubuntu:22.04 (compiling and now also without SIGILL during start, https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/1607607257).

LynxAbraxas committed:
Reapply "switching from debian -> ubuntu:22.04 ...":

as SIGILL problem is solved with 4c412e64:
see https://github.com/civctp2/civctp2/issues/417#issuecomment-2559334624

build.log now contains warnings on missing returns

This reverts commit 5ac8059c7de1e12f4d4f2a27dbde103d5ee6464f.

see also 4ca7c9b3

LynxAbraxas committed:
avoid configure error on guessing build types by using --build arg of configure:

https://stackoverflow.com/questions/4810996/how-to-resolve-configure-guessing-build-type-failure#68025766

LynxAbraxas committed:
avoid configure error on guessing build types by using newer config.guess and config.sub (should work for various arch):

partially reverts commit 94c8661ccf3583d425d0b84a7fe1d15fd4c1064e
https://stackoverflow.com/questions/4810996/how-to-resolve-configure-guessing-build-type-failure#6786522

LynxAbraxas commented:
And with some modifications also successfully configures for ubuntu:24.10 and now also yields warnings on missing returns:
https://gitlab.com/LynxAbraxas/civctp2/-/jobs/8748336961#L10864

LynxAbraxas committed:
add std-err also to build.log

LynxAbraxas committed:
plain shell does not know |&, i.e. redirect stderr to stdout before piping:

https://stackoverflow.com/questions/35917552/what-does-the-syntax-mean-in-shell-language#comment59539400_35917682

LynxAbraxas committed:
build-in diff-rules for *.h and *.cpp files

LynxAbraxas committed:
base DF on ubuntu:24.04 (was ubuntu:rolling till Apr. 2025)

LynxAbraxas committed:
base DF on ubuntu:rolling (ubuntu:25.04, ubuntu:plucky), to have a moving DI-tag to a stable distro:

https://hub.docker.com/_/ubuntu

see https://github.com/civctp2/civctp2/issues/417#issuecomment-2559334624

This reverts commit 2a71b4c72573a26ef56418c6aa4ff65eb003be55.

see also 4ca7c9b3

LynxAbraxas committed:
new configure switch for activating ffmpeg-code for movies, disabled by default because the FFmpeg-API still changes a lot

LynxAbraxas committed:
build plain DI image without ffmpeg if building with ffmpeg fails (debug DI without ffmpeg)

MartinGuehmann commented:
> Seems the last tests of main with #435 failed (https://gitlab.com/civctp2/civctp2/-/pipelines/1987707574) due to debian:latest having changed to a newer release, which would also need some changed contained in this PR (e.g. [bd5da69](https://github.com/civctp2/civctp2/commit/bd5da693fa2f00431383b0b1738d42aa1d1e1a26)) for newer ubuntu images. With the changes of this PR for ubuntu:22.04 compilation of #435 again works (https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/2000148838), however halve of the smoke tests fail.

Which I find quite strange that it says that the c compiler is not working. But anyway, thanks.
 
> This PR is still WIP, because I cannot get the SDL_ffmpeg code to work with the new ffmpeg libraries contained in ubuntu:rolling (ubuntu:25.04, ubuntu:plucky), e.g. libavcodec61 based on ffmpeg_7.1.1. The code of [9bb7809](https://github.com/civctp2/civctp2/commit/9bb78090552143300d8dae66220da9056f487e51) compiles (https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/2016002112) but ctp2 crashes as soon as a video should be played without any error on stderr or the logs, even for the debug version. @MartinGuehmann @williamd67 do you have any idea what could still be wrong/missing?

On Windows, I had a bug that made the game hang when it tried to load a file that was not there: d03c148c17fd5a4cba1516ea2a9354a5fda0a3f4. But I guess that is unreleated.

LynxAbraxas committed:
downgrade docker to avoid API version problems:

https://support.gitlab.com/hc/en-us/articles/23529961961372-Client-version-1-xx-is-too-old-Minimum-supported-API-version-is-1-44-please-upgrade-your-client-to-a-newer-version-error-in-Docker

LynxAbraxas committed:
FindPath expects float for cost instead of double (yields error for newer gcc, reverts part of 6da70fc086)

MartinGuehmann commented:
> The code of [9bb7809](https://github.com/civctp2/civctp2/commit/9bb78090552143300d8dae66220da9056f487e51) compiles (https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/2016002112) but ctp2 crashes as soon as a video should be played without any error on stderr or the logs, even for the debug version.

You can try again with merging in #437. There, I implemented the generating a crash.txt on Linux, it will also print something to the terminal. It also does this if you end the game with Ctrl+c from the terminal, might be useful for endless loops.

The other thing is for #437 I also cherry picked bd5da693fa2f00431383b0b1738d42aa1d1e1a26, but there seems also be another problem added in the meantime.

LynxAbraxas committed:
changes on MD5 code needed for gcc >=14 and -std=gnu23:

taken from https://github.com/edoapra/nwchem/commit/8ad88697c00bd7eb7db6436980a31662e8c6fe7d
see: https://github.com/nwchemgit/nwchem/issues/1082

possible ToDo: change to newer sources for MD5: https://docs.openssl.org/master/man3/MD5/

LynxAbraxas committed:
additional dependency for CTPDatabase.cpp to avoid build errors due to missing DB-code

LynxAbraxas committed:
do not retry build stages multiple times to avoid loss of CI-minutes:

possibly not needed any more with e7aa50aa: "additional dependency for CTPDatabase.cpp to avoid build errors due to missing DB-code"

default runners from GL.com have multiple vCPUs:
https://docs.gitlab.com/ci/runners/hosted_runners/linux/#machine-types-available-for-linux---x86-64

LynxAbraxas commented:
> > The code of [9bb7809](https://github.com/civctp2/civctp2/commit/9bb78090552143300d8dae66220da9056f487e51) compiles (https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/2016002112) but ctp2 crashes as soon as a video should be played without any error on stderr or the logs, even for the debug version.
> 
> You can try again with merging in #437. There, I implemented the generating a crash.txt on Linux, it will also print something to the terminal. It also does this if you end the game with Ctrl+c from the terminal, might be useful for endless loops.

OK, will check that later. Currently, trying to get GL-CI compile with ffmpeg first and if that fails without ffmepg (see nee commits here). Then the porting of the ffmpeg code to the newer API can go in its own PR.

> The other thing is for #437 I also cherry picked [bd5da69](https://github.com/civctp2/civctp2/commit/bd5da693fa2f00431383b0b1738d42aa1d1e1a26), but there seems also be another problem added in the meantime.

Possibly that of [8b0150e](https://github.com/civctp2/civctp2/pull/428/commits/8b0150e311d8fbb7d8762c441396b2ac2aeb4e24) or any other of the new commits here?

LynxAbraxas committed:
disable compilation with ffmpeg for plain as well (GL-CI optional stage not working, to be moved to a new PR)

LynxAbraxas committed:
do not error on implicit-function-declaration (needed with newer gcc) to avoid error on:

scyaccdefs.h:45:15: error: implicit declaration of function 'sclex' [-Wimplicit-function-declaration]
   45 | #define yylex sclex
      |               ^~~~~
sc.tab.c:215:16: note: in expansion of macro 'yylex'
  215 | # define YYLEX yylex()
      |                ^~~~~
sc.tab.c:1004:18: note: in expansion of macro 'YYLEX'
 1004 |         yychar = YYLEX;
      |                  ^~~~~

possibly related to:
https://stackoverflow.com/questions/70016128/lexyacc-program-gives-error-including-implicit-declaration-of-yylex

LynxAbraxas commented:
rebased PR onto current master, compiles fine without ffmpeg (https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/2236311186) but the smoke test on the sprite editor fail with a segfault on loading e.g. gg006.spr:
https://gitlab.com/LynxAbraxas/civctp2/-/jobs/12559426035#L106
@MartinGuehmann could that originate from some of your latest changes?
It seems to not get the file name `gg006.spr`any more: https://gitlab.com/LynxAbraxas/civctp2/-/jobs/12559426035#L103

I also observe a `double free or corruption (!prev)` when quitting the game.

LynxAbraxas commented:
Could come from https://github.com/civctp2/civctp2/commit/77a5f64026a629a1aa560620c0ee3bf4ea580af6#diff-3d497b3b1ae112fa4d835608e0ecc23111cedce6d9426ea9c0389e27735533fbR545

MartinGuehmann commented:
> Could come from [77a5f64#diff-3d497b3b1ae112fa4d835608e0ecc23111cedce6d9426ea9c0389e27735533fbR545](https://github.com/civctp2/civctp2/commit/77a5f64026a629a1aa560620c0ee3bf4ea580af6#diff-3d497b3b1ae112fa4d835608e0ecc23111cedce6d9426ea9c0389e27735533fbR545)

These are just changes to the format strings. The sprite group stuff should have blown up directly into my face. Than you have the one in SpriteEditor that just tells sscanf to interpret the number in the sprite script file as a short int (sint16) instead off an int (sint32). There seems nothing wrong with that.

MartinGuehmann commented:
> rebased PR onto current master, compiles fine without ffmpeg (https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/2236311186) but the smoke test on the sprite editor fail with a segfault on loading e.g. gg006.spr: https://gitlab.com/LynxAbraxas/civctp2/-/jobs/12559426035#L106 @MartinGuehmann could that originate from some of your latest changes? It seems to not get the file name `gg006.spr`any more: https://gitlab.com/LynxAbraxas/civctp2/-/jobs/12559426035#L103

I don't see that it does not get the file name rather it crashes when it tries to load that. Actually, what it does there exactly. I just loaded the debug version and part of the SpriteEditor user interface is black. And I get a lot of Asserts, when I click on Load. Obviously, this is not how it should be.

Unfortunately, the stack trace is not generated properly, and so is useless.

> I also observe a `double free or corruption (!prev)` when quitting the game.

I have not seen that, but I think I still have to check that.

MartinGuehmann commented:
That's the stack trace; I get when I hit the load button in the SpriteEditor:

```
Windows x86 Version 2025-12-29
Exception: 'Access Violation' thrown.
  0x0130e088  [??_R4MessageOpenAction@@6B@ + 0x0]
  0x006c6079  [?ColorBlt@aui_Blitter@@UAE?AW4AUI_ERRCODE@@PAVaui_Surface@@PAUtagRECT@@KI@Z + 0x149]
  0x0078b838  [?ReDrawLargeSprite@SpriteEditWindow@@QAEXXZ + 0x78]
  0x0078aae3  [?Idle@SpriteEditWindow@@UAE?AW4AUI_ERRCODE@@XZ + 0x103]
  0x006ece13  [?Idle@aui_UI@@UAE?AW4AUI_ERRCODE@@PAVaui_Region@@@Z + 0x33]
  0x006edb6f  [?Process@aui_UI@@UAE?AW4AUI_ERRCODE@@XZ + 0xf]
  0x007d2dab  [?Process@CivApp@@QAEHXZ + 0x1bb]
  0x007cae2f  [?CivMain@@YGHPAUHINSTANCE__@@0PADH@Z + 0x30f]
  0x007ce3de  [WinMain@16 + 0x6e]
  0x00fa6ae5  [?__scrt_common_main_seh@@YAHXZ + 0xf8]
  0x757bfcc9  [$R002030 + 0x72e0caa9]
  0x774a82ae  [$R002030 + 0x74af508e]
  0x774a827e  [$R002030 + 0x74af505e]
```

Win32 SDL version. The top one looks a bit odd.

LynxAbraxas committed:
clean up LoadSprite to avoid segfault when loading from text field

LynxAbraxas commented:
OK, with e26d6bf3a1297f7a255 the segfault from the spriteEditor is gone and all test pass again:
https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/2237708348

I decided to disable compilation with ffmpeg for movies (f7ba61d9c43420fe8b47c83bc08497af04173c1a) moving that into a new PR #440

@MartinGuehmann what do you think? Ready for merging?

LynxAbraxas committed:
smoke tests: click on text field twice because sometimes first click has no effect

LynxAbraxas committed:
disable optimization which causes spurious segfaults, double free and other crashes (tested -O1, -O2, -O3):

optimization was not applied till f465fc9ef22ca3587d4ff2db436d820f89094896
see also 759183fba04b473469e1a33f4932902bcd0704a4 of #95

LynxAbraxas commented:
With 70379053bca406fb7da914de7f4290cc4884f1cf  the spurious segfaults etc (discussed here: https://github.com/civctp2/civctp2/pull/437#issuecomment-3701598494) seem to have disappeared. It seems the current code base is not suitable for optimization by the compiler.

MartinGuehmann commented:
I would prefer to to have --disable-ffmpeg4movies than --enable-ffmpeg4movies, or at least have it mentioned in the readme. I would also prefer it this way for the deb file. Since, the executable from the deb file is broken right now anyway, it doesn't matter right now.

Since, the spurious seq faults have nothing to do with the optimization, I would go back to -O3. And also mention the problem in the readme.

But i can also do the readme things after merging, since i consider it more important to have this in.

LynxAbraxas commented:
Still investigating the spurious crashes. Currently I suspect it is a change in SDL-mixer from 2.0.4 (which seems to work all right) till 2.8.1, possibly concerning OGG/Vorbis.

MartinGuehmann commented:
That would mean the spurious crashes go away if you disable the music. Either in game or by removing the move files from their folder.

MartinGuehmann commented:
No real problems with that, so can go in.
